### PR TITLE
ResourceCollection should not have a path

### PR DESCRIPTION
**Pull request by gregw**
A `ResourceCollection` should have a null path and URI.  It is by definition a collection of different paths and URIs, so it cannot just pick one of them and return it.   This is hiding lots of bugs in the code where `resource.getPath()` is called and assumes it will always get a non-null value that holds all the required resources.  However, when scanning for descriptors, annotations, TLDs etc. the underlying resource is often a collection, thus code that uses getPath is currently only scanning the first resource in the collection.

This PR fixes `ResourceCollection`, but will almost certainly break some other tests.   I think we want to use other PRs to fix those uses of `getPath` first before considering completing this PR.

### CODE REVIEW

**Comment by joakime** [comment]
If we go this path, then I think `ResourceCollection` should also not have  ...

* `lastModified()`
* `length()`
* `getURI()`
* `getFileName()`
* `getName()`

The changes you are doing here also apply to `MemoryResource` btw

**Comment by joakime** [comment]
The code overall would be so much easier with a `List<>` instead of hiding this collection behind an interface that is for a single entry.

**Comment by gregw** [comment]
> If we go this path, then I think `ResourceCollection` should also not have ...
> 
> * `lastModified()`

lastModified has a sensible and meaningful interpretation for composite directories: the most recently modified time of all the composites

> * `length()`

Agreed.

> * `getURI()`

Yep, should return null... I think this PR already does.

> * `getFileName()`

I'm on the fence about this one.... if for example you have a composite base resource and resolve "WEB-INF", then you may get back a composite that contains `WEB-INF` directories from multiple resources in the base composite.   In that case, since they are all called `WEB-INF`, I think it is reasonable to return "WEB-INF" for the filename.    If they don't agree, then return null. That's how I have implemented it for now.

Flip side, is that if it always returned null, I don't think it would be a big problem... except perhaps for debugging.... maybe when we are naming contexts etc.   So I think on balance going with the impl in this PR is OK.

> * `getName()`

Yep.

> The changes you are doing here also apply to `MemoryResource` btw

Oh I thought we got rid of that class?  Is it still used? If not, let's kill it.    If we are still using it, then I'll also review.

**Comment by gregw** [comment]
> The code overall would be so much easier with a `List<>` instead of hiding this collection behind an interface that is for a single entry.

Absolutely not!  

We have a bunch of methods that can apply to a composite collection of resources that we have to think carefully about how to interpret them against a collection: is it the first; is it a combination; is it null; is it allowed at all?    With a `ResourceCollection` class we can put all those difficult decisions in one place and make them consistent.  We can police the creation of the collection to only be non-null directory resources.

If we pass around a `List<Resource>` instead, then
 - every usage of that list will have to make the difficult choices:  is it the first; is it a combination; is it null; is it allowed at all?  Inevitably different choices will be made in different places so we will have inconsistency... a file might be visible for annotation scanning but not for TLD scanning!
- We will have duplicated code iterating over the lists in multiple places... or perhaps we decide to have static utility methods into which we pass the list? In which case we have just re-invented object oriented programming.
- Since we could not enforce the construction of the `List<Resource>` to contained only non-null directory resources, then we would need to spread sanity and null checking code everywhere.  We'd need to check on every usage of the collection rather than just check on the creation of the collection.  So instead of doing those checks once, we could end up doing them hundreds of thousands of times per second on every request.
- A directory is itself just a collection of other resources and we have many places where a single such collection is simple to deal with (e.g. scanning WEB-INF).  Turning that simple abstraction of a collection of files into a collection of collections of files is needlessly complex. 
- The readability of the code will be reduced.  Instead of `web-inf.lastModified()` we will have `Instance last = null; for (Resource r : list) { Instant l = r.lastModified(); if (last == null || l.isAfter(last)) last = l; } return last;`... repeated.... in many places.... ?!?!?!?

Note that there are valid usages for `List<Resource>` and we have many examples in the code.   We might have multiple descriptors, or multiple jar files or multiple webapps etc.   In these cases, the list is not trying to represent a single resource, but is genuinely representing multiple resources that the caller needs to explicitly handle in whatever capacity it is doing.  These are not use-cases for `ResourceCollection` because they are not abstracting a single resource.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 1]
```diff
@@ -230,7 +230,10 @@ public long length()
      */
     public InputStream newInputStream() throws IOException
     {
-        return Files.newInputStream(getPath(), StandardOpenOption.READ);
+        Path path = getPath();
+        if (path == null)
+            return null;
```
Comment (lachlan-roberts) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:205]: I would expect this to throw `IOException` not return null from the Javadoc. 
So I think the javadoc should be updated as well.

**Comment by joakime** [comment]
> > If we go this path, then I think `ResourceCollection` should also not have ...
> > 
> > * `lastModified()`
> 
> lastModified has a sensible and meaningful interpretation for composite directories: the most recently modified time of all the composites

I'm not sure that's a valid use case, where do you see this being used like this?

> > * `length()`
> 
> Agreed.
> 
> > * `getURI()`
> 
> Yep, should return null... I think this PR already does.
> 
> > * `getFileName()`
> 
> I'm on the fence about this one.... if for example you have a composite base resource and resolve "WEB-INF", then you may get back a composite that contains `WEB-INF` directories from multiple resources in the base composite. In that case, since they are all called `WEB-INF`, I think it is reasonable to return "WEB-INF" for the filename. If they don't agree, then return null. That's how I have implemented it for now.

Interesting, I'll have to see if we use this from a Directory based Resource.

> Flip side, is that if it always returned null, I don't think it would be a big problem... except perhaps for debugging.... maybe when we are naming contexts etc. So I think on balance going with the impl in this PR is OK.

Most debugging should be using either `getName()` or `getURI()` as it's more clear.

> > * `getName()`
> 
> Yep.
> 
> > The changes you are doing here also apply to `MemoryResource` btw
> 
> Oh I thought we got rid of that class? Is it still used? If not, let's kill it. If we are still using it, then I'll also review.

The original proposal was in (the now closed, un-merged) PR #8470

Now, the work Lachlan is doing for HttpContent.Factory can easily get rid of MemoryResource, as it can serve a static HttpContent for `jetty-dir.css` as a fallback.
Lets wait for that work to complete and it'll be easy to remove MemoryResource.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java [hunk 2]
```diff
@@ -112,7 +113,7 @@ public interface Context extends ClassVisibilityChecker
          */
         boolean isParentLoaderPriority();
 
-        ResourceCollection getExtraClasspath();
+        Resource getExtraClasspath();
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java]: @lorban @joakime @sbordet  I believe it was a miss-use of `ResourceCollection` when this was changed from a `List<Resource>`.    We should very VERY rarely ever explicitly use a `ResourceCollection`, because the whole point of that class is to look like a singe resource.   If anything needs to iterate over the contained resources, then it probably should be using a `List<Resources>` rather than a `ResourceCollection`.

However, for not, I have only partially reverted to try how it looks if extra classpath is just a `Resource`, which may or may not be a collection.   This requires the `addClassPath` method below to be somewhat `ResourceCollection` away, as it needs to extract URLs.  I've done this for now with a static stream method.... but not sure I like.   Reverting all the way back to `List<Resource>` may be the best thing to do here.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java [hunk 3]
```diff
@@ -223,23 +219,30 @@ public Context getContext()
...
-     * @param resources The resources to add to the classpath
-     * @throws IOException if unable to add classpath from resource
+     * @param resource The resources to add to the classpath
      */
-    public void addClassPath(Resource resources)
-        throws IOException
+    public void addClassPath(Resource resource)
     {
-        for (Resource resource: resources)
+        ResourceCollection.stream(resource).forEach(r ->
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java]: Since any `Resource` might be a collection, if we need URLs, URIs or Paths from a resource, then we should use this stream abstraction to get back the one or more values.  The stream can then be used to pick the first, collect into a list or do forEach handling like I am below:

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java (hunk 2 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java]: maybe `ResourceCollection` should be renamed to `MultiResource` to emphasise that it is-a `Resource` and should not be thought of as a `Collection` (which is an implementation detail).

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java [hunk 4]
```diff
@@ -35,13 +35,15 @@
     /**
      * <p>Make a Resource containing a collection of other resources</p>
      * @param resources multiple resources to combine as a single resource. Typically, they are directories.
-     * @return A Resource of multiple resources.
+     * @return A Resource of multiple resources or a single resource if only 1 is passed, or null if none are passed
      * @see ResourceCollection
      */
-    static ResourceCollection combine(List<Resource> resources)
+    static Resource combine(List<Resource> resources)
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:41]: `combine` should return a `Resource` and not a `ResourceCollection`, since if the passed list has only a single entry, then a collection is not needed.    The caller should not need to know if it is a collection or not.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java (hunk 2 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppClassLoader.java]: Better name options include:
 + `MultiResource`
 + `CompositeDirectoryResource`
 + `CombinedResource`
 + `CombinedDirectoryResource`

**Comment by gregw** [comment]
@joakime nudge

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorProcessor.java [hunk 5]
```diff
@@ -110,7 +110,7 @@ public void process(Request request, Response response, Callback callback)
 
             try
             {
-                generateAcceptableResponse(request, response, code, message, cause, callback);
+                generateResponse(request, response, code, message, cause, callback);
```
Comment (joakime) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorProcessor.java]: Unrelated to this PR.
Plus, this has "acceptable" in it's name as it's generating based on the `Accept` header.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ReHandlingErrorProcessor.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.ExceptionUtil;
+import org.eclipse.jetty.util.URIUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An ErrorProcessor that can re-handle a request at an error page location.
+ */
+public abstract class ReHandlingErrorProcessor extends ErrorProcessor
```
Comment (joakime) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ReHandlingErrorProcessor.java]: Also unrelated to the scope of this PR,
Can you pull this out into a different PR?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/AttributeNormalizer.java [hunk 7]
```diff
@@ -11,7 +11,7 @@
 // ========================================================================
 //
 
-package org.eclipse.jetty.ee9.quickstart;
+package org.eclipse.jetty.util.resource;
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/AttributeNormalizer.java:14]: Different PR for this too

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java [hunk 8]
```diff
@@ -26,36 +23,52 @@
...
 
 import org.eclipse.jetty.util.URIUtil;
 
 /**
- * A collection of Resources.
- * Allows webapps to have multiple sources.
- * The first resource in the collection is the main resource.
- * If a resource is not found in the main resource, it looks it up in
- * the order the provided in the constructors
+ * Multiple resource directories presenteda as a single Resource.
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java]: Typo `presented as`

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java [hunk 9]
```diff
@@ -26,36 +23,52 @@
...
+public class CombinedResource extends Resource
 {
-    private final List<Resource> _resources;
+    /**
+     * Stream a resource to its component resources (if any).
+     * @param resource The resource to stream
+     * @return If the passed resource is a ResourceCollection, then it's contained
+     *         resources are streamed, otherwise the passed resource itself if streamed
+     */
+    public static Stream<Resource> stream(Resource resource)
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java]: Wrong place for this.
Perhaps this should be in `Resource` or `Resources` ?

Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java]: Why not just have `.stream()` on the `Resource.stream()` interface instead?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java [hunk 10]
```diff
@@ -26,36 +23,52 @@
...
-     * Instantiates a new resource collection.
-     *
-     * @param resources the resources to be added to collection
+     * <p>Make a Resource containing a combination of other resources</p>
+     * @param resources multiple resources to combine as a single resource. Typically, they are directories.
+     * @return A Resource of multiple resources or a single resource if only 1 is passed, or null if none are passed
+     * @see CombinedResource
      */
-    ResourceCollection(List<Resource> resources)
+    static Resource combine(List<Resource> resources)
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:40]: Isn't this on `ResourceFactory.combine()` already?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java [hunk 11]
```diff
@@ -127,127 +153,67 @@ public Resource resolve(String subUriPath)
...
-            }
-        }
         return null;
     }
 
     @Override
     public String getFileName()
     {
+        String filename = null;
+        // return a non-null filename only if all resources agree on the same name.
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:186]: Perhaps we should make `getFileName()` not be the last segment name in case of Directory?   Is there a use case were in we care about, and use `getFileName()` in a directory case?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 12]
```diff
@@ -198,22 +198,28 @@ public long length()
...
     }
 
     /**
      * Readable ByteChannel for the resource.
      *
-     * @return an readable bytechannel to the resource or null if one is not available.
+     * @return a readable {@link java.nio.channels.ByteChannel} to the resource or null if one is not available.
      * @throws IOException if unable to open the readable bytechannel for the resource.
      */
     public ReadableByteChannel newReadableByteChannel() throws IOException
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:215]: This should probably be a `SeekableByteChannel` instead.
(Would be more generally useful in the static byte-range support as well)

**Review by joakime — CHANGES_REQUESTED** [review]
My not-yet-fully-reviewed review

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 12 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:215]: The `Files.newByteChannel()` already returns `SeekableByteChannel`

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java]: The style we have been adopting in jetty-12 is to avoid adding methods to interfaces if they can be implemented in a static like this.

However, this is not necessarily the correct place for this static.... and there is already the iterator API that does the same thing.... so I will review....

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 10 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:40]: It is, which calls this version, which has access to the private `gatherUniqueFlatResourceList` method which then calls the private constructor, that no longer does the sanity checks, as they are not needed when creating a CombinedResource from resolve.   So this allows us to avoid needless iterations and checks.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 11 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:186]: We can use the fileName to generate a context name.  I'm not 100% sure we still use this method for that, but it is a reasonable use-case to ask a directory for the name of it's last segment.

This method exists and I think this is a reasonable interpretation of it for a combined resource.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 12 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:215]: probably... I'll do that in another PR.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/AttributeNormalizer.java (hunk 7 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/AttributeNormalizer.java:14]: Actually this change is for this PR.  It turns out that the `AttributeNormalizer` has for a while been assuming that a Path is available, so the main change here is to make it use the stream API for CombinedResources to generate multiple URI to resource mappings.

But as the Normalizer code was identical in ee9 and ee10, I also moved to resource to avoid duplicating the change. 

So I think the change does belong here.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 10 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:40]: Note that this method  is only package scope, so it is not public API.  It is just to get the implementation in the right file rather than in ResourceFactory.java

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 12 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:215]: Opened as PR #8772

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java [hunk 13]
```diff
@@ -30,32 +27,34 @@
...
-     * Instantiates a new resource collection.
-     *
-     * @param resources the resources to be added to collection
+     * <p>Make a Resource containing a combination of other resources</p>
+     * @param resources multiple resources to combine as a single resource. Typically, they are directories.
+     * @return A Resource of multiple resources or a single resource if only 1 is passed, or null if none are passed
+     * @see CombinedResource
      */
-    ResourceCollection(List<Resource> resources)
+    static Resource combine(List<Resource> resources)
```
Comment (lachlan-roberts) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:40]: Should this also have a `Resource combine(Resource... resources)` same as the `ResourceFactory`?

File: jetty-ee10/jetty-ee10-maven-plugin/src/test/java/org/eclipse/jetty/ee10/maven/plugin/TestWebAppPropertyConverter.java [hunk 14]
```diff
@@ -152,10 +152,10 @@ public void testFromProperties() throws Exception
         assertEquals(true, webApp.isPersistTempDirectory());
         assertEquals(war.getAbsolutePath(), webApp.getWar());
         assertEquals(webXml.getAbsolutePath(), webApp.getDescriptor());
-        assertThat(webApp.getBaseResource(), instanceOf(ResourceCollection.class));
+        assertThat(webApp.getBaseResource(), instanceOf(CombinedResource.class));
 
-        ResourceCollection resourceCollection = (ResourceCollection)webApp.getBaseResource();
-        List<URI> actual = resourceCollection.getResources().stream().filter(Objects::nonNull).map(Resource::getURI).toList();
+        CombinedResource combinedResource = (CombinedResource)webApp.getBaseResource();
+        List<URI> actual = combinedResource.getResources().stream().filter(Objects::nonNull).map(Resource::getURI).toList();
```
Comment (lachlan-roberts) [inline jetty-ee10/jetty-ee10-maven-plugin/src/test/java/org/eclipse/jetty/ee10/maven/plugin/TestWebAppPropertyConverter.java]: Probably better to use `Resource.stream()`.

**Review by lachlan-roberts — APPROVED** [review]
Looks good to me but there are some test failures.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ReHandlingErrorProcessor.java (hunk 6 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ReHandlingErrorProcessor.java]: synced with head.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java]: moved to Resource

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java (hunk 13 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:40]: It's not a public API.  It is just there to move the implementation into the correct file.   So if the ... is not needed, no need to add it.
